subsys/fs/fs.c

```diff
@@ -255,11 +255,25 @@ int fs_readdir(struct fs_dir_t *zdp, struct fs_dirent *entry)
 		int rc = -EINVAL;
 
 		if (zdp->mp->fs->readdir != NULL) {
-			rc = zdp->mp->fs->readdir(zdp, entry);
+			/* Loop until error or not special directory */
+			while (true) {
+				rc = zdp->mp->fs->readdir(zdp, entry);
+				if (rc < 0) {
+					break;
+				}
+				if (entry->type != FS_DIR_ENTRY_DIR) {
+					break;
```

Comment:
Shoudn't this case returns an errcode?

Reply:
No.  If it's not a directory entry, it's a candidate for return (probably a file, maybe something else).  This just allows fall through to the subsequent check for names of special directories, the last condition for whether something gets filtered out.